### ISSUE TITLE
InceptionResnetV1 last layers

### ISSUE THREAD

**Issue by nosound2** [timesler/facenet-pytorch#55]
Hi Tim, thank you for the great code.

I try to make sense of the output with `classify = True` of `InceptionResnetV1` and things do not sum up for me. So I looked at the TF code that you reference, and I have a couple of question, can you please help me to figure that out

1. The code that I am looking at is [this for the model](https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py) and [this for the training and the classification layer](https://github.com/davidsandberg/facenet/blob/master/src/train_softmax.py). Why do you use `self.last_bn`, it is not mentioned in the TF code. I assume every `slim.fully_connected` comes with a batch norm (where is it defined?), but how do you copy the weights then? Some black magic is happenning :)

2. You call `x = F.normalize(x, p=2, dim=1)` unconditionally, but it seems like the TF version only does `tf.nn.l2_normalize` for embeddings, which are basically not used afterwards. I suspect there is a bug on your side, specifically in the [finetuning notebook](https://github.com/timesler/facenet-pytorch/blob/master/examples/finetune.ipynb) you apply `torch.nn.CrossEntropyLoss()` with that normalized activations, while the TF version does not do it. But in general `x = self.logits(x)` operation should be applied without normalization (again, according to the TF version). 

Thanks,
Zahar

**Comment by timesler** [comment]
@nosound2 thanks for raising this.

> 1. The code that I am looking at is [this for the model](https://github.com/davidsandberg/facenet/blob/master/src/models/inception_resnet_v1.py) and [this for the training and the classification layer](https://github.com/davidsandberg/facenet/blob/master/src/train_softmax.py). Why do you use `self.last_bn`, it is not mentioned in the TF code. I assume every `slim.fully_connected` comes with a batch norm (where is it defined?), but how do you copy the weights then? Some black magic is happenning :)

You are right to assume that `slim.fully_connected` includes batch norm. It is included via a context manager in the davd sandberg repo (https://github.com/davidsandberg/facenet/blob/096ed770f163957c1e56efa7feeb194773920f6e/src/models/inception_resnet_v1.py#L146). A great example of overly convoluted and difficult to understand tensorflow code :)

> 2. You call x = F.normalize(x, p=2, dim=1) unconditionally, but it seems like the TF version only does tf.nn.l2_normalize for embeddings, which are basically not used afterwards. I suspect there is a bug on your side, specifically in the finetuning notebook you apply torch.nn.CrossEntropyLoss() with that normalized activations, while the TF version does not do it. But in general x = self.logits(x) operation should be applied without normalization (again, according to the TF version).

You are correct - I've fixed this bug in master and the latest version of the pip package
